[BUG] Layout engine much slower on Chrome

When working with complex layouts the layout engine sometimes takes a considerable time to rerender because it has to measure the size of various components. For a long time I thought this was a general issue with the layout engine but it was just pointed out to me that this is **much** worse in Chrome.

In the contrived example below we have a reasonably complex layout of rows and columns and keep appending new rows with some text and plots.

```python
from bokeh.layouts import Row, Column, Spacer
from bokeh.models.widgets import Div, Slider, Button
from bokeh.plotting import figure
from bokeh.io import curdoc

title = Div(text='<h1>Page title</h1>')

header = Row(children=[title, Spacer(sizing_mode='stretch_width'),
                       Div(text="Some link")],
             height=60, sizing_mode="stretch_width")

main = Column(sizing_mode='stretch_width')

def on_click(event):
    p = figure(sizing_mode='stretch_width')
    p.line([1, 2, 3], [1, 2, 3])
    row = Row(children=[Div(text="Some text") for i in range(10)]+[p],
              sizing_mode='stretch_width')
    main.children.append(row)

widgets = [Slider(start=0, end=10, title='Slider %d' % i) for i in range(10)]
button = Button(label='Click me')
button.on_click(on_click)
widgets.append(button)

nav = Column(children=widgets, width=200)

layout = Column(children=[header, Row(children=[nav, main])])

curdoc().add_root(layout)
```

The layout engine struggles the more items you add with the button presses in Chrome, while in Firefox it hardly slows down. Here are flame graphs generated by recording the performance in the two browsers.

## Chrome

<img width="1121" alt="Screen Shot 2019-12-08 at 2 43 10 PM" src="https://user-images.githubusercontent.com/1550771/70390325-43cfc700-19ca-11ea-8880-10cb743f41be.png">

## Firefox

<img width="956" alt="Screen Shot 2019-12-08 at 2 54 30 PM" src="https://user-images.githubusercontent.com/1550771/70390377-b2148980-19ca-11ea-8247-266d23d53ac2.png">

You will note that it takes <2 seconds in Firefox and ~5 in Chrome. I'm not sure if this is just a Chrome performance issue but it would at least be good to investigate why it's so much slower and if anything can be done to fix it.

Versions:

Bokeh 1.4.0
Chrome Version 78.0.3904.108 (Official Build) (64-bit)
Firefox 70.0.1 (64-bit)

And just for comparison. It's much faster in a Jupyter Notebook. But Firefox is significantly faster than Chrome.

![notebook_performance](https://user-images.githubusercontent.com/42288570/70494190-ab8d2b80-1b0a-11ea-82cb-445ad7ec90f3.gif)


Im running this notebook https://github.com/MarcSkovMadsen/awesome-panel/blob/master/app.ipynb

and have this log

```bash
$ jupyter notebook app.ipynb
[I 04:54:31.256 NotebookApp] Serving notebooks from local directory: C:\repos\private\awesome-panel
[I 04:54:31.256 NotebookApp] The Jupyter Notebook is running at:
[I 04:54:31.257 NotebookApp] http://localhost:8888/?token=5f809af8188893692b9f863c8db8159841e4768130891117
[I 04:54:31.257 NotebookApp]  or http://127.0.0.1:8888/?token=5f809af8188893692b9f863c8db8159841e4768130891117
[I 04:54:31.257 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 04:54:31.891 NotebookApp]

    To access the notebook, open this file in a browser:
        file:///C:/Users/masma/AppData/Roaming/jupyter/runtime/nbserver-7816-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/?token=5f809af8188893692b9f863c8db8159841e4768130891117
     or http://127.0.0.1:8888/?token=5f809af8188893692b9f863c8db8159841e4768130891117
[I 04:54:36.293 NotebookApp] Kernel started: 0b477488-b6e8-433c-be95-42b46e49a125
[W 04:54:41.925 NotebookApp] 404 GET /api/kernels/6ed7af11-6e24-4a43-9c3f-b873b8db42db/channels?session_id=5ac847a8828f46bc8e3e2b7fa8e5955f (::1): Kernel does not exist: 6ed7af11-6e24-4a43-9c3f-b873b8db42db
[W 04:54:41.939 NotebookApp] 404 GET /api/kernels/6ed7af11-6e24-4a43-9c3f-b873b8db42db/channels?session_id=5ac847a8828f46bc8e3e2b7fa8e5955f (::1) 16.01ms referer=None
[I 04:54:46.510 NotebookApp] 302 GET /?token=5f809af8188893692b9f863c8db8159841e4768130891117[I%2004:54:31.257%20NotebookApp]%20Use%20Control-C%20to%20stop%20this%20server%20and%20shut%20down%20all%20kernels%20(twice%20to%20skip%20confir (127.0.0.1) 2.00ms
[I 04:54:46.584 NotebookApp] Saving file at /app.ipynb
[W 04:55:14.917 NotebookApp] Replacing stale connection: 6ed7af11-6e24-4a43-9c3f-b873b8db42db:5ac847a8828f46bc8e3e2b7fa8e5955f
[I 04:55:41.064 NotebookApp] Starting buffering for 0b477488-b6e8-433c-be95-42b46e49a125:fc57efbb9e6b40c3ab18bb4aec6e49f4
[I 04:55:42.401 NotebookApp] Kernel shutdown: 0b477488-b6e8-433c-be95-42b46e49a125
[I 04:55:45.803 NotebookApp] Kernel started: 4c1f0636-7239-43ae-b32b-d7c6a6b7fd19
[I 04:57:46.056 NotebookApp] Saving file at /app.ipynb
[W 05:00:19.905 NotebookApp] Replacing stale connection: 6ed7af11-6e24-4a43-9c3f-b873b8db42db:5ac847a8828f46bc8e3e2b7fa8e5955f
```

and this error message

```javascript
Javascript error adding output!
Error: Mismatched anonymous define() module: function(){return function(){return function t(e,r,n){function a(o,s){if(!r[o]){if(!e[o]){var l="function"==typeof require&&require;if(!s&&l)return l(o,!0);if(i)return i(o,!0);var c=new Error("Cannot find module '"+o+"'");throw c.code="MODULE_NOT_FOUND",c}var u=r[o]={exports:{}};e[o][0].call(u.exports,function(t){return a(e[o][1][t]||t)},u,u.exports,t,e,r,n)}return r[o].exports}for(var i="function"==typeof require&&require,o=0;o<n.length;o++)a(n[o]);return a}}()({1:[function(t,e,r){"use strict";var n=t("../src/lib"),a={"X,X div":"direction:ltr;font-family:'Open Sans', verdana, arial, sans-serif;margin:0;padding:0;","X input,X button":"font-family:'Open Sans', verdana, arial, sans-serif;","X input:focus,X button:focus":"outline:none;","X a":"text-decoration:none;","X a:hover":"text-decoration:none;","X .crisp":"shape-rendering:crispEdges;","X .user-select-none":"-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none;","X svg":"overflow:hidden;","X svg a":"fill:#447adb;","X svg a:hover":"fill:#3c6dc5;","X .main-svg":"position:absolute;top:0;left:0;pointer-events:none;","X .main-svg .draglayer":"pointer-events:all;","X
....
....
....
```

--- a/bokehjs/src/lib/core/layout/html.ts
+++ b/bokehjs/src/lib/core/layout/html.ts
@@ -16,7 +16,6 @@ export class ContentBox extends ContentLayoutable {
 }
 
 export class VariadicBox extends Layoutable {
-
   constructor(readonly el: HTMLElement) {
     super()
   }
@@ -28,5 +27,29 @@ export class VariadicBox extends Layoutable {
       const {border, padding} = extents(this.el)
       return content.grow_by(border).grow_by(padding).map(Math.ceil)
     })
+  }
+}
+
+export class CachedVariadicBox extends VariadicBox {
+  private readonly _cache: Map<string, SizeHint> = new Map()
+
+  constructor(el: HTMLElement) {
+    super(el)
+  }
+
+  protected _measure(viewport: Size): SizeHint {
+    const {width, height} = viewport
+    const key = `${width},${height}`
+
+    let size_hint = this._cache.get(key)
+    if (size_hint == null) {
+      size_hint = super._measure(viewport)
+      this._cache.set(key, size_hint)
+    }
+    return size_hint
+  }
+
+  invalidate_cache(): void {
+    this._cache.clear()
   }
 }

--- a/bokehjs/src/lib/models/widgets/markup.ts
+++ b/bokehjs/src/lib/models/widgets/markup.ts
@@ -1,4 +1,4 @@
-import {VariadicBox} from "core/layout"
+import {CachedVariadicBox} from "core/layout/html"
 import {div} from "core/dom"
 import * as p from "core/properties"
 
@@ -9,12 +9,14 @@ import clearfix_css from "styles/clearfix.css"
 
 export abstract class MarkupView extends WidgetView {
   model: Markup
+  layout: CachedVariadicBox
 
   protected markup_el: HTMLElement
 
   connect_signals(): void {
     super.connect_signals()
     this.connect(this.model.change, () => {
+      this.layout.invalidate_cache()
       this.render()
       this.root.compute_layout() // XXX: invalidate_layout?
     })
@@ -25,7 +27,7 @@ export abstract class MarkupView extends WidgetView {
   }
 
   _update_layout(): void {
-    this.layout = new VariadicBox(this.el)
+    this.layout = new CachedVariadicBox(this.el)
     this.layout.set_sizing(this.box_sizing())
   }
 

